### PR TITLE
Implement 3DGUT

### PR DESCRIPTION
**Pull request by liruilong940607**
This PR implements the paper

"3DGUT: Enabling Distorted Cameras and Secondary Rays in Gaussian Splatting"

which allows gsplat to support: 
- **distorted** pinhole/opencv camera 
- **rolling shutter** effects for both training and inference.

https://github.com/user-attachments/assets/291481ec-9546-4d50-a737-19422dbadffd

The PR is carefully tested & benchmarked.[See below]

### CODE REVIEW

**Comment by liruilong940607** [comment]
MCMC 3DGS 1M on MipNeRF360 at 30k  (`benchmark/mcmc.sh`):

| Scene   | PSNR  | SSIM   | LPIPS | Train Mem (GB) | Train Time (Sec) |
|---------|-------|--------|-------|----------------|------------------|
| garden  | 27.32 | 0.8574 | 0.104 | 1.30           | 882              |
| bicycle | 25.62 | 0.7765 | 0.185 | 1.32           | 861              |
| stump   | 26.96 | 0.7907 | 0.164 | 1.31           | 860              |
| bonsai  | 32.73 | 0.9539 | 0.106 | 1.38           | 1226             |
| counter | 29.48 | 0.9241 | 0.130 | 1.49           | 1393             |
| kitchen | 31.58 | 0.9359 | 0.087 | 1.39           | 1263             |
| room    | 32.40 | 0.9380 | 0.132 | 1.43           | 1235             |

MCMC 3DGUT 1M on MipNeRF360 at 30k  (`benchmark/3dgut/mcmc.sh`):

| Scene   | PSNR  | SSIM   | LPIPS | Train Mem (GB) | Train Time (Sec) |
|---------|-------|--------|-------|----------------|------------------|
| garden  | 27.20 | 0.8566 | 0.105 | 1.30           | 1114             |
| bicycle | 25.52 | 0.7739 | 0.184 | 1.32           | 1105             |
| stump   | 26.77 | 0.7876 | 0.167 | 1.31           | 1079             |
| bonsai  | 32.65 | 0.9536 | 0.106 | 1.38           | 1663             |
| counter | 29.46 | 0.9238 | 0.129 | 1.47           | 2076             |
| kitchen | 31.50 | 0.9362 | 0.087 | 1.39           | 1829             |
| room    | 32.32 | 0.9376 | 0.133 | 1.40           | 1714             |

**Comment by liruilong940607** [comment]
MCMC FisheyeGS 1M on ZipNeRF Dataset at 30k (`benchmark/fisheye/mcmc_zipnerf.sh`):

| Scene   | PSNR  | SSIM   | LPIPS | Train Mem (GB) | Train Time (Sec) |
|---------|-------|--------|-------|----------------|------------------|
| nyc     | 26.41 | 0.9004 | 0.109 | 2.74           | 2722             |
| alameda | 20.57 | 0.8065 | 0.178 | 3.23           | 3668             |
| berlin  | 24.85 | 0.9046 | 0.166 | 3.33           | 4167             |
| london  | 25.60 | 0.8521 | 0.154 | 3.40           | 3964             |

MCMC 3DGUT 1M on ZipNeRF Dataset at 30k (`benchmark/3dgut/mcmc_zipnerf.sh`):

| Scene   | PSNR  | SSIM   | LPIPS | Train Mem (GB) | Train Time (Sec) |
|---------|-------|--------|-------|----------------|------------------|
| nyc     | 26.67 | 0.9032 | 0.108 | 2.74           | 2890             |
| alameda | 20.52 | 0.8068 | 0.178 | 3.22           | 3846             |
| berlin  | 24.86 | 0.9041 | 0.168 | 3.33           | 4579             |
| london  | 25.71 | 0.8551 | 0.153 | 3.39           | 4136             |

File: README.md [hunk 1]
```diff
@@ -11,6 +11,10 @@ gsplat is an open-source library for CUDA accelerated rasterization of gaussians
   <video src="https://github.com/nerfstudio-project/gsplat/assets/10151885/64c2e9ca-a9a6-4c7e-8d6f-47eeacd15159" width="100%" />
 </div>
 
+## News
+
+[April 2025] [NVIDIA 3DGUT](https://research.nvidia.com/labs/toronto-ai/3DGUT/) is not integrated in gsplat! Checkout [here](docs/3dgut.md) for more details.
```
Comment (zgojcic) [inline README.md:16]: is not -> is now

**Comment by gradeeterna** [comment]
Wow, excited to try this! Please could we also get support for separate binary masks in gsplat? This was recently added to the official 3DGUT so can use circular fisheyes which require masking out around the image circle. Thanks!

**Comment by gradeeterna** [comment]
@liruilong940607 Ah cool, I remember with FisheyeGS it automatically masked the corners when transforming Zipnerf full frame fisheyes to idealized. How would I use this with circular fisheyes and separate binary masks? Due to the 180 degree fov, I also need to mask out myself from a lot of images as well as the black edges. 

image
mask

**Comment by gradeeterna** [comment]
My datasets are in opencv_fisheye format and I can run them fine with FisheyeGS in gsplat and other methods. I just can't figure out to how use my own masks in gsplat, which mask out the black areas and also myself from the back facing lens.

**Comment by ishipachev** [comment]
@gradeeterna 
You can prepare masks in colmap-type format (if you work with its DB), duplicate it for each image and use colmap data loader. But be aware, that it may slow down training.
```ns-train lalala colmap --masks-path path-to-masks"

**Comment by scott198510** [comment]
> This PR implements the paper
> 
> "3DGUT: Enabling Distorted Cameras and Secondary Rays in Gaussian Splatting"
> 
> which allows gsplat to support:
> 
> * **distorted** pinhole/opencv camera
> * **rolling shutter** effects for both training and inference.
> 
>  gsplat-3dgut.mp4 
> The PR is carefully tested & benchmarked.[See below]

@liruilong940607 

After reading the updated version of the code, I found that it is not as you described—directly inputting radially distorted images with distortion for training and rendering. In fact, it still performs distortion removal, just like the original version.
#
#
        # Load data: Training data should contain initial points and colors.
        self.parser = Parser(
            data_dir=cfg.data_dir,
            factor=cfg.data_factor,
            normalize=cfg.normalize_world_space,
            test_every=cfg.test_every,
        )

            if camtype == "perspective":
                K_undist, roi_undist = cv2.getOptimalNewCameraMatrix(
                    K, params, (width, height), 0
                )
                mapx, mapy = cv2.initUndistortRectifyMap(
                    K, params, None, K_undist, (width, height), cv2.CV_32FC1
                )
                mask = None

As mentioned in Paper 3DGUT, the original images are directly used for training and rendering, and the output retains the same intrinsic parameters as the input. If you first undistort the images and then use them for training, you effectively alter the original intrinsic parameters, making it impossible to restore the images to their original form. This is because undistorting a pinhole camera's radially distorted images, for example, inevitably leads to the loss of peripheral pixels.

**Comment by liruilong940607** [comment]
@scott198510 

Yes you are right the benchmark script you are looking at in the main branch first undistort the images into a perfect pinhole/fisheye. This is just to align with the fisheyeGS paper's train/val setting so we can compare.

The gsplat rasterization API, however, does support passing in distortion parameters directly. The script for benchmarking on the raw distorted images can be found on another branch here [use the same gsplat rasterization API with the main branch just updated the training script and dataloader a bit]:

https://github.com/nerfstudio-project/gsplat/blob/dev/perfect_camera/examples/benchmarks/3dgut/mcmc_zipnerf_with_distortion.sh

**Comment by scott198510** [comment]
> @scott198510
> 
> Yes you are right the benchmark script you are looking at in the main branch first undistort the images into a perfect pinhole/fisheye. This is just to align with the fisheyeGS paper's train/val setting so we can compare.
> 
> The gsplat rasterization API, however, does support passing in distortion parameters directly. The script for benchmarking on the raw distorted images can be found on another branch here [use the same gsplat rasterization API with the main branch just updated the training script and dataloader a bit]:
> 
> https://github.com/nerfstudio-project/gsplat/blob/dev/perfect_camera/examples/benchmarks/3dgut/mcmc_zipnerf_with_distortion.sh

@liruilong940607 
I checked the corresponding code in the branch you mentioned, and I'm very disappointed—it only supports fisheye. I just wanted to see how to train with a pinhole camera model that includes radial distortion. Paper 3DGUT mentioned this approach.


        if not self.parser.process_to_perfect_camera:
            # pass out distortion parameters so that we could consider
            # distortion during rendering
            assert self.parser.camtype == "fisheye", (
                f"Only fisheye camera is supported for now, got {self.parser.camtype}"
            )
            radial_coeffs = params[:4]
            data["radial_coeffs"] = torch.from_numpy(radial_coeffs).float()

**Comment by liruilong940607** [comment]
@scott198510 I see we didn't provide that particular demo but it's very easily doable. Just take a look at how we do it for fisheye, i believe you don't need to change more than 20 lines to achieve it. [gsplat's rasterization API already fully supports distorted pinhole camera, you just need to pass the distortion parameters from dataset to the rasterization API function]

We didnt provide such training script because 1/ I was lazy 2/ we were benchmarking against other papers which are all train/eval on undistorted camera]

However we did provide a rendering script to demonstrate how you can pass in pinhole/fisheye distortion parameters to gsplat's API: 

https://github.com/nerfstudio-project/gsplat/blob/main/examples/simple_viewer_3dgut.py#L199-L201

We also carefully examine the gradient of the function to make sure it works for training as well [although we didn't provide such a training script] [The announcement on "gsplat supports xxx"  refers to the core gsplat API, not the training script. The training script -- as well as the render script -- at this moment is just to demonstrate how to use gsplat's API]

**Comment by scott198510** [comment]
> @scott198510 I see we didn't provide that particular demo but it's very easily doable. Just take a look at how we do it for fisheye, i believe you don't need to change more than 20 lines to achieve it. [gsplat's rasterization API already fully supports distorted pinhole camera, you just need to pass the distortion parameters from dataset to the rasterization API function]
> 
> We didnt provide such training script because 1/ I was lazy 2/ we were benchmarking against other papers which are all train/eval on undistorted camera]
> 
> However we did provide a rendering script to demonstrate how you can pass in pinhole/fisheye distortion parameters to gsplat's API:
> 
> https://github.com/nerfstudio-project/gsplat/blob/main/examples/simple_viewer_3dgut.py#L199-L201
> 
> We also carefully examine the gradient of the function to make sure it works for training as well [although we didn't provide such a training script] [The announcement on "gsplat supports xxx" refers to the core gsplat API, not the training script. The training script -- as well as the render script -- at this moment is just to demonstrate how to use gsplat's API]
@liruilong940607 
Thanks! I will do it.
I noticed your replies all came within about an hour - truly a Chinese/Chinese diaspora working style. Truly delighted!
